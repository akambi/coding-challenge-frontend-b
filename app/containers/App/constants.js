export const QUEBEC = 'quebec';
export const MONTREAL = 'montreal';
export const API_KEY = 'PARTNER_BaASYYHxTxuOINEOMWq5GA';

export const CityToGeoHashMapping = {
  quebec: 'f2m673',
  montreal: 'f25dvk',
};

export const CHANGE_SEARCH_PARAMS = 'busbud/App/CHANGE_SEARCH_PARAMS';

export const LOAD_DEPARTURES = 'busbud/App/LOAD_DEPARTURES';
export const LOAD_DEPARTURES_SUCCESS = 'busbud/App/LOAD_DEPARTURES_SUCCESS';
export const LOAD_DEPARTURES_ERROR = 'busbud/App/LOAD_DEPARTURES_ERROR';
