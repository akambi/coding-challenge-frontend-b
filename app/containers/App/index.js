import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import Footer from 'components/Footer';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Busbud's bus departures"
        defaultTitle="Busbud's bus departures"
      >
        <meta
          name="description"
          content="A simple application that lists all departures for a given origin city and a given destination city for a given day"
        />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}
