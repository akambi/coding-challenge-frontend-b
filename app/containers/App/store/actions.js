/*
 * App Actions
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_SEARCH_PARAMS,
  LOAD_DEPARTURES,
  LOAD_DEPARTURES_SUCCESS,
  LOAD_DEPARTURES_ERROR,
} from '../constants';

/**
 * Changes the search parameters
 *
 * @param  {object} searchParams The new search parameters
 *
 * @return {object} An action object with a type of CHANGE_SEARCH_PARAMS
 */
export function changeSearchParams(searchParams) {
  return {
    type: CHANGE_SEARCH_PARAMS,
    searchParams,
  };
}

/**
 * Load the departures, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_DEPARTURES
 */
export function loadDepartures(poll = false) {
  return {
    type: LOAD_DEPARTURES,
    poll,
  };
}

/**
 * Dispatched when the departures are loaded by the request saga
 *
 * @param  {array} departures The departure data
 *
 * @return {object}      An action object with a type of LOAD_DEPARTURES_SUCCESS passing the departures
 */
export function departuresLoaded(departures) {
  return {
    type: LOAD_DEPARTURES_SUCCESS,
    departures,
  };
}

/**
 * Dispatched when loading the departures fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_DEPARTURES_ERROR passing the error
 */
export function departureLoadingError(error) {
  return {
    type: LOAD_DEPARTURES_ERROR,
    error,
  };
}
