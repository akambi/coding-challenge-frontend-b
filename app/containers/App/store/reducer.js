/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  CHANGE_SEARCH_PARAMS,
  LOAD_DEPARTURES_SUCCESS,
  LOAD_DEPARTURES,
  LOAD_DEPARTURES_ERROR,
  QUEBEC,
  MONTREAL,
} from '../constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  searchParams: {
    origin: QUEBEC,
    destination: MONTREAL,
    outboundDate: new Date().toISOString(),
    adult: 1,
  },
  searchResult: {
    departures: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_SEARCH_PARAMS:
        draft.searchParams = action.searchParams;
        break;

      case LOAD_DEPARTURES:
        draft.loading = true;
        draft.error = false;
        draft.searchResult.departures = false;
        break;

      case LOAD_DEPARTURES_SUCCESS:
        draft.searchResult.departures = action.departures;
        draft.loading = false;
        break;

      case LOAD_DEPARTURES_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
