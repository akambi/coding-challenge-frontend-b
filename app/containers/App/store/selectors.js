/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectSearchParams = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.searchParams,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectDepartures = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.searchResult.departures,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectSearchParams,
  makeSelectDepartures,
  makeSelectLoading,
  makeSelectError,
  makeSelectLocation,
};
