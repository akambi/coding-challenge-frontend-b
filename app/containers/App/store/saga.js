/**
 * Gets all departures for a given origin city and a given destination city for a given day
 */

import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import {
  API_KEY,
  LOAD_DEPARTURES,
  CityToGeoHashMapping,
} from 'containers/App/constants';
import {
  loadDepartures,
  departuresLoaded,
  departureLoadingError,
} from 'containers/App/store/actions';

import request from 'utils/request';
import { makeSelectSearchParams } from './selectors';

/**
 * Busbud search departures request/response handler
 */
export function* getDepartures(action) {
  const headers = new Headers({
    Accept:
      'application/vnd.busbud+json; version=2; profile=https://schema.busbud.com/v2/',
    'X-Busbud-Token': API_KEY,
  });

  const options = {
    method: 'GET',
    headers,
    mode: 'cors',
    cache: 'default',
  };

  // Select search params from store
  const searchParams = yield select(makeSelectSearchParams());

  const origin = CityToGeoHashMapping[searchParams.origin];
  const destination = CityToGeoHashMapping[searchParams.destination];
  const date = new Date(searchParams.outboundDate).toISOString().split('T')[0];
  const { adult } = searchParams;

  let requestURL = `https://napi.busbud.com/x-departures/${origin}/${destination}/${date}?adult=${adult}`;
  if (action && action.poll) {
    requestURL = `https://napi.busbud.com/x-departures/${origin}/${destination}/${date}/poll?adult=${adult}`;
  }

  try {
    const result = yield call(request, requestURL, options);

    if (!result.complete) {
      // sleep 2 seconds
      yield delay(2 * 1000);
      yield put(loadDepartures(true));
      return;
    }

    const departures = result.departures.map(departure => ({
      departure_time: departure.departure_time,
      arrival_time: departure.arrival_time,
      price: departure.prices.total,
      location: result.locations.find(
        location => location.id === departure.origin_location_id,
      ),
    }));

    yield put(departuresLoaded(departures));
  } catch (err) {
    yield put(departureLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* busbudData() {
  // Watches for LOAD_DEPARTURES actions and calls getDepartures when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_DEPARTURES, getDepartures);
}
