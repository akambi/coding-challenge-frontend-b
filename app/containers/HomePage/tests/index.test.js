/**
 * Test the HomePage
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { HomePage, mapDispatchToProps } from '../index';
import { changeSearchParams, loadDepartures } from '../../App/store/actions';
import configureStore from '../../../configureStore';
import { MONTREAL, QUEBEC } from '../../App/constants';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore(
      {
        global: {
          searchParams: {
            origin: QUEBEC,
            destination: MONTREAL,
            outboundDate: '2020-12-14T14:21:30.490Z',
            adult: 1,
          },
        },
      },
      browserHistory,
    );
  });

  beforeEach(() => {
    window.Headers = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const searchParams = {
      origin: QUEBEC,
      destination: MONTREAL,
      outboundDate: '2020-12-14T14:21:30.490Z',
      adult: 1,
    };
    const submitSpy = jest.fn();
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage
            loading={false}
            error={false}
            departures={[]}
            searchParams={searchParams}
            onSubmitForm={submitSpy}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch the departures on mount', () => {
    const submitSpy = jest.fn();
    const searchParams = {
      origin: 'quebec',
      destination: 'montreal',
      outboundDate: '2020-12-13',
      adult: 1,
    };

    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage
            searchParams={searchParams}
            onchangeSearchParams={() => {}}
            onSubmitForm={submitSpy}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch loadDepartures when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const searchParams = {
          origin: 'quebec',
          destination: 'montreal',
          outboundDate: '2020-12-13',
          adult: 1,
        };

        result.onSubmitForm(searchParams);
        expect(dispatch).toHaveBeenCalledWith(changeSearchParams(searchParams));
        expect(dispatch).toHaveBeenCalledWith(loadDepartures());
      });

      it('should not dispatch changeSearchParams and loadDepartures when searchParams is not passed', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onSubmitForm();
        expect(dispatch).not.toHaveBeenCalledWith(changeSearchParams(null));
        expect(dispatch).not.toHaveBeenCalledWith(loadDepartures());
      });
    });
  });
});
