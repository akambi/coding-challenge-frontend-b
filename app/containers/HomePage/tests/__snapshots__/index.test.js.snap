// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<article>
  <div>
    <section
      class="Section-yrd82d-0 CenteredSection-sc-9nc1xs-0 dHSliU"
    >
      <h2
        class="H2-jhzvdt-0 jcQAUa"
      >
        <span>
          Search all departures for a given origin city and a given destination city for a given day
        </span>
      </h2>
    </section>
    <section
      class="Section-yrd82d-0 SearchSection-sc-9353ax-0 fqcZqV"
    >
      <div
        class="search-params-form-container"
      >
        <form
          class="search-params-form"
        >
          <div
            class="search-origin"
          >
            <label
              for="origin"
            >
              <span>
                Origin
              </span>
            </label>
            <select
              class="Select-qu3sxg-0 fumbYO"
              id="origin"
            >
              <option
                value="quebec"
              >
                Québec
              </option>
            </select>
          </div>
          <div
            class="search-destination"
          >
            <label
              for="destination"
            >
              <span>
                Destination
              </span>
            </label>
            <select
              class="Select-qu3sxg-0 fumbYO"
              id="destination"
            >
              <option
                value="montreal"
              >
                Montréal
              </option>
            </select>
          </div>
          <div
            class="search-outbound-date"
          >
            <label
              for="outboundDate"
            >
              <span>
                Date
              </span>
            </label>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class=""
                  id="outboundDate"
                  type="text"
                  value="12/14/2020"
                />
              </div>
            </div>
          </div>
          <div
            class="search-adult"
          >
            <label
              for="adult"
            >
              <span>
                Adults
              </span>
            </label>
            <input
              id="adult"
              min="1"
              type="number"
              value="1"
            />
          </div>
          <div
            class="submit"
          >
            <button
              class="submit-btn StyledButton-e3glmp-0 iVzJYl"
              type="submit"
            >
              <span>
                Search
              </span>
            </button>
          </div>
        </form>
      </div>
    </section>
    <section
      class="Section-yrd82d-0 ldfMvH"
    >
      <div
        class="Wrapper-sc-1l7ow4n-0 gikgZI"
      >
        <ul
          class="Ul-sc-1l9y7is-0 jTPOp"
        />
      </div>
    </section>
  </div>
</article>
`;
